sdk/typescript/src/types/framework.ts

```diff
@@ -57,18 +62,113 @@ export class Coin {
     };
   }
 
-  static getBalance(
-    data: GetObjectDataResponse | SuiMoveObject
-  ): BN | undefined {
+  public static getID(obj: ObjectData): ObjectId {
+    if (isSuiMoveObject(obj)) {
+      return obj.fields.id.id;
+    }
+    return getObjectId(obj);
+  }
+
+  /**
+   * Convenience method for select coin objects that has a balance greater than or equal to `amount`
+   *
+   * @param amount coin balance
+   * @param exclude object ids of the coins to exclude
+   * @return a list of coin objects that has balance greater than `amount` in an ascending order
+   */
+  static selectCoinsWithBalanceGreaterThanOrEqual(
+    coins: ObjectDataFull[],
+    amount: bigint,
+    exclude: ObjectId[] = []
+  ): ObjectDataFull[] {
+    return Coin.sortByBalance(
+      coins.filter(
+        (c) => !exclude.includes(Coin.getID(c)) && Coin.getBalance(c)! >= amount
+      )
+    );
+  }
+
+  /**
+   * Convenience method for select a minimal set of coin objects that has a balance greater than
+   * or equal to `amount`. The output can be used for `PayTransaction`
+   *
+   * @param amount coin balance
+   * @param exclude object ids of the coins to exclude
+   * @return a minimal list of coin objects that has a combined balance greater than or equal
+   * to`amount` in an ascending order. If no such set exists, an empty list is returned
+   */
+  static selectCoinSetWithCombinedBalanceGreaterThanOrEqual(
+    coins: ObjectDataFull[],
+    amount: bigint,
+    exclude: ObjectId[] = []
+  ): ObjectDataFull[] {
+    const sortedCoins = Coin.sortByBalance(
+      coins.filter((c) => !exclude.includes(Coin.getID(c)))
+    );
+
+    const total = Coin.totalBalance(sortedCoins);
+    // return empty set if the aggregate balance of all coins is smaller than amount
+    if (total < amount) {
+      return [];
+    } else if (total === amount) {
+      return sortedCoins;
+    }
+
+    return Coin.sortByBalance(Coin.selectCoinSetRecursive(sortedCoins, amount));
```

Comment:
iirc, the main point of doing this recursively is that, when the left amount can be covered by 
```
const coinWithSmallestSufficientBalance = sortedCoins.find(
      (c) => Coin.getBalance(c)! >= amount
    );
```
instead of the biggest one left, imo this might not be worth the cost of having recursion here; instead we can just iterate here until hitting the total amount needed.

The reasons that the impact of picking different coins is minimal to me are:
- all the new Pay related transactions will take a list of coins - and for the Sui related ones, it can take many dusts of coins and handle gas payment as well; 
- our gas selection cannot optimize for all, some ppl might want to avoid dusts for some reason and thus want to start with smallest coins, thus here I prefer to have a lightweight impl. which gives us a working solution for the majority of use cases. (sorting itself is probably okay though)

Reply:
> imo this might not be worth the cost of having recursion here

The only reason to use recursion here is I though Javascript runtime supports tail recursion optimization, but it seems that's not the case. I've converted the original code to iterative.